### ISSUE TITLE
Aparapi should default to CPU mode before JTP mode if OpenCL successfully generated

### ISSUE THREAD

**Issue by GoogleCodeExporter** [tigerneil/aparapi#16]
```
During our testing of Aparapi, we've encountered a number of instances where 
the OpenCL code has been successfully created, but fails to execute on the 
targeted GPU.

But we have also noticed that if we force Aparapi to use CPU mode instead of 
JTP mode, the OpenCL code has generally executed 2x or greater performance on 
CPU than JTP.

Since OpenCL is intended to support GPU/CPU/FPGA with the same code, we would 
like to request the following change:

(pseudo code)

if opencl_generated_successfully
{
  try GPU
if GPU fails
  try CPU
if CPU fails
  try JTP
}
else
{
 try JTP
}
```

Original issue reported on code.google.com by `ryan.lam...@gmail.com` on 10 Nov 2011 at 6:13

**Comment by GoogleCodeExporter** [comment]
```
This is a good suggestion (it came up somewhere else recently, I thought it was 
in another issue thread but a search was fruitless).  

Not all workloads txfer as well to the CPU, although the CPU is generally at 
least comparable to JTP.  I have seen it worse only a few times, I do see some 
workloads that benefit greatly (although 2x is impressive - I would like to see 
more details on the workload).

Of course not all OpenCL runtimes support CPU offload (NVidia for example does 
not), and of course not all runtimes support GPU (Intel).  So I suspect the 
pseudocode would be 

if opencl_available {
   if opencl_generated_successfully
     if gpu available
        try GPU
        if GPU fails{
           if cpu available{ 
              try CPU
              if CPU fails{
                 JTP
              }
           }else{
              JTP
           }
        }else{
           JTP
        }
     }else{
        JTP
     }
   }else{
     JTP
   }
}else{
   JTP
}

Maybe the trick is to arrange this fallback strategy as you suggest and also 
allow the fallback path to be set in the execution mode.  Maybe adding an 
explicit varargs setExecutionFallbackPath() so we can express our preference. 

Kernel.setExecutionFallbackPath(GPU, JTP); // will try GPU if available and 
default to JTP

Kernel.setExecutionFallbackPath(CPU, JTP); // will not even try GPU, will try 
CPU if available and default to JTP

Kernel.setExecutionFallbackPath(GPU); // will try GPU but will fail if GPU not 
available (no automatic JTP)

What do we think of this.

```

Original comment by `frost.g...@gmail.com` on 10 Nov 2011 at 6:51
- Changed state: **Accepted**
- Added labels: **Type-Enhancement**
- Removed labels: **Type-Defect**

**Comment by GoogleCodeExporter** [comment]
```
I like that idea a lot and would give us a lot of flexibility based on our 
different environments.

It would also be important to make sure that the fallback code in Aparapi is as 
robust as possible, so that any failures during these steps would at least 
guarantee execution at some level (except the last option, as noted).
```

Original comment by `ryan.lam...@gmail.com` on 10 Nov 2011 at 7:01

**Comment by GoogleCodeExporter** [comment]
```
I just found the other thread where this came up.  For reference it was issue 
#9. The user expected fallback route through CPU. 

We will track the enhancement request here.
```

Original comment by `frost.g...@gmail.com` on 10 Nov 2011 at 7:22

**Comment by GoogleCodeExporter** [comment]
```
Issue 9 has been merged into this issue.
```

Original comment by `frost.g...@gmail.com` on 10 Nov 2011 at 7:24

**Comment by GoogleCodeExporter** [comment]
```
Steven Libby provided a patch for this. Tested and merged into trunk (r#627) 
8/21. 

I will ask Steven to create a Wiki page to describe the new fallback strategy.
```

Original comment by `frost.g...@gmail.com` on 21 Aug 2012 at 4:10
- Changed state: **Fixed**
